### PR TITLE
fstab-generator: disable default deps if x-systemd.{wanted,required}-by= is used

### LINKED ISSUE #30273: x-systemd.wanted-by documentation and implementation do not match
### systemd version the issue has been seen with

main

### Used distribution

_No response_

### Linux kernel version used

_No response_

### CPU architectures issue was seen on

None

### Component

other

### Expected behaviour you didn't see

The documentation says:

x-systemd.wanted-by=, x-systemd.required-by=

In the created mount unit, configures a WantedBy= or RequiredBy= dependency on another unit. This option may be specified more than once. If this is specified, the normal automatic dependencies on the created mount unit, e.g., local-fs.target, are not automatically created. See WantedBy= and RequiredBy= in [systemd.unit(5)](https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#) for details.

### Unexpected behaviour you saw

In the source code we can see that the Before dependency is only controlled by the nofail option.

https://github.com/systemd/systemd/blob/d4506301f945006c1619f7455b35653517b57a74/src/fstab-generator/fstab-generator.c#L604

### Steps to reproduce the problem

1. Add x-systemd.wanted-by=default.target to a fstab line
2. There's no change to how local-fs.target is rendered to the generated .mount file

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by YHNdnzj**
Fixes #30273

### CODE REVIEW

**Comment by YHNdnzj** [comment]
> Is it a regression?

It's not AFAICS. The problem has existed since the introduction of the options (be02c1cf4).

So, feel free to postpone this. We can backport later if necessary.

**Comment by bluca** [comment]
Given it's very late I'm wary of changing dependencies handling at boot, so I'd prefer to postpone

**Comment by keszybz** [comment]
I agree. **Any** change to dependency logic has potential for catastrophic failure.

**Review by bluca — APPROVED** [review]

**Comment by YHNdnzj** [comment]
The new implementation matches the documentation though. If people use this, they ought to configure the possible After= dependencies themselves. The original request was exactly about dropping the ordering deps?

**Comment by YHNdnzj** [comment]
> Also, this potentially breaks existing setups.

Yes, but the risk should be low enough IMO. In most cases, the option is used when you want to specify a target that's different from e.g. local-fs.target. So the ordering deps makes little sense.

**Comment by yuwata** [comment]
> The new implementation matches the documentation

Oh, really? I did not know that. I will re-review later.

File: man/systemd.mount.xml
Comment (yuwata) [inline man/systemd.mount.xml:249]: > The new implementation matches the documentation

Indeed...

File: man/systemd.mount.xml
Comment (yuwata) [inline man/systemd.mount.xml:150]: Out of scope of this PR, and not your fault, but this section does not mention about the default dependency for tmpfs mount.

File: man/systemd.mount.xml
Comment (yuwata) [inline man/systemd.mount.xml]: I think it is better to mention that it may be better to also specify x-systemd.after or x-systemd.before, as the default ordering dependencies will be dropped if x-systemd.{wanted,required}-by is specified.

**Review by yuwata — APPROVED** [review]
OK. Then, LGTM.
But, I'd like to wait for @poettering's final review.

**Comment by YHNdnzj** [comment]
Made the suggested changes and added one more commit. PTAL.

**Comment by yuwata** [comment]
@poettering Please take a final look.

**Pull request merged**
